util/src/main/java/com/redhat/lightblue/util/metrics/DropwizardRequestMetrics.java

```diff
@@ -0,0 +1,145 @@
+
+ You should have received a copy of the GNU General Public License
+ along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package com.redhat.lightblue.util.metrics;
+
+import static com.codahale.metrics.MetricRegistry.name;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.List;
+import java.util.Objects;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.UndeclaredThrowableException;
+
+import com.redhat.lightblue.util.Error;
+import com.codahale.metrics.Counter;
+import com.codahale.metrics.Timer;
+import com.codahale.metrics.MetricRegistry;
+
+public class DropwizardRequestMetrics implements RequestMetrics {
+    
+    private final Logger LOGGER = LoggerFactory.getLogger(DropwizardRequestMetrics.class);
+
+    private static final String API = "api";
+
+    private final MetricRegistry metricsRegistry;
+    
+    // replace any special characters in error message with standard metric separator
+    private final String REGEX = "[-@#!$%^&*()_+|~={}:;'<>?,\"\\s]";
+
+    public DropwizardRequestMetrics(MetricRegistry metricRegistry) {
+        metricsRegistry = metricRegistry;
+    }
+
+    /**
+     * Create exception namespace for metric reporting based on exception name
+     * 
+     */
+    private static String errorNamespace(String metricNamespace, Throwable exception, String errorMessage) {
+        Class<? extends Throwable> actualExceptionClass = unravelReflectionExceptions(exception);
+        return name(metricNamespace, "requests", "exception", actualExceptionClass.getSimpleName(), errorMessage);
+    }
+
+    /**
+     * Get to the cause we actually care about in case the bubbled up exception is a
+     * higher level framework exception that encapsulates the stuff we really care
+     * about.
+     * 
+     */
+    private static Class<? extends Throwable> unravelReflectionExceptions(Throwable e) {
+        while (e.getCause() != null
+                && (e instanceof UndeclaredThrowableException || e instanceof InvocationTargetException)) {
+            e = e.getCause();
+        }
+        return e.getClass();
+    }
+
+    @Override
+    public Context startEntityRequest(String operation, String entity, String version) {
+        return new DropwizardContext(name(API, operation, entity, version));
+    }
+
+    @Override
+    public Context startStreamingEntityRequest(String operation, String entity, String version) {
+        return new DropwizardContext(name(API, "stream", operation, entity, version));
+    }
+    
+    @Override
+    public Context startLockRequest(String lockOperation, String domain) {
+        return new DropwizardContext(name(API, "lock", domain, lockOperation));
+    }
+
+    @Override
+    public Context startSavedSearchRequest(String searchName, String entity, String version) {
+        return new DropwizardContext(name(API, "savedsearch", searchName, entity, version));
+    }
```

Comment:
For all of these that use version, if version is null, this will roll up to to just "operation.entity" which means technically we lose the ability to distinguish between default version searches and other versions. How about if version is null, using the string "default" instead?

Reply:
Added a null check to replace null version with default as suggested.

Reply:
Thanks!

We found one more thing that we need to address with version. Basically, right now the convention is that versions use semver like "1.0.0". However "." is a meaningful character with graphite metrics, so this creates 3 path segments "1", "0", and "0" where the default version will only have one (just "default"). When you do queries with graphite, you use wild cards per path segment, per the docs:

> All wildcards apply only within a single path element. In other words, they do not include or cross dots (.). Therefore, servers.* will not match servers.ix02ehssvc04v.cpu.total.user, while servers.*.*.*.* will.

So that means it'd be pretty difficult if not impossible to have a query look at all versions including default.

The solution I believe is rather simple: let's just replace all of the "." with some other character like "_" in version. Then both default and explicit versions will use only one path segment.

Reply:
Replaced "." with "_" in version. Also removed the regex pattern used earlier to format error code with a "." separated value and replaced it with a replace(":","-") . So now the error "mongo-crud:DatabaseError" will be reported as "mongo-crud-DatabaseError" (a single entity for easier graphite query), instead of mongo.crud.DatabaseError.

The only reason I included regex in first place was because having some special characters in metric namespace caused it to display a bit differently in visualvm. For eg:- 
without special characters - api.find.user.1.2.3.requests.active 
with special characters - "api.find.user.1.2.3.requests.exception.Error.mongo-metadata:UnknownVersion" (notice the opening and closing quote)

While doing the version changes, found that the quotes get added to namespace only for a few special characters and not for all.So if we have "_"  or " - " in namespace, it displays normaly  and only ":" was causing issues. 
Since ":" is the only other special character we use in our error codes apart from "-", replacing just this one shou;ld suffice for our needs.